A couple of optimizations

vrurg opened the pull request:
Add method `ASSIGN-KEY` and re-write `AT-KEY` to get rid of `Proxy` object.

vrurg committed:
Don't use Proxy

Add method `ASSIGN-KEY` and re-write `AT-KEY` to get rid of `Proxy`
object.

vrurg commented:
I believe this should be more efficient implementation.

lizmat commented:
The problem is that `AT-KEY` is supposed to return a container for hashes.  Aka, this should work;
````raku
my %h;
%h.AT-KEY("a") = 42;
dd %h  # Hash %h = {:a(42)}
````

vrurg commented:
Then `Proxy` could be an alternative branch, a fallback.

lizmat commented:
I'm afraid I'm not following...

vrurg committed:
Make AT-KEY method writable again

`%ordered.AT-KEY($key) = $value`

vrurg committed:
Use delegation for basic methods

This should make run-time inlining easier since less stack frames would
be involved. Besides, while `handles` is not currently statically
optimized, it could be in the future.

vrurg commented:
See the updates. `Proxy` would only be involved for missing keys.

As a matter of fact, I don't think a 3rd party module should strictly follow the specs. Having a non-writable `AT-KEY` wouldn't bother me as long as there is working `ASSIGN-KEY`. 

Yet, often used `%h<missing-key> // do {...}` construct would suffer from proxy penalty. Would it be possible to determine wether `AT-KEY` is directly invoked or via a postcircumfix op then it could make a big difference. But...

lizmat commented:
Closing as I think the Proxy issue is unsolvable, and the PR has bitrotted.

Thanks for trying though!
